### PR TITLE
Issue #143 + some pep8

### PR DESCRIPTION
**Pull request by takacsg84**

### CODE REVIEW

**Comment by f0k** [comment]
Hey, thank you, but that does a little bit too much at once. Could you please just address #143 here and leave the batch normalization for another PR?
Also the pep8 cleanup in `updates.py` will conflict with the pep8 cleanup in #135, so you will need to rebase after #135 is merged.
(If you need any help with the necessary git juggling, let me know!)

**Comment by benanne** [comment]
Yeah, better to split these things up.

From a quick look at the code (I haven't looked at in detail yet) I feel like it could be a lot shorter.  We've also been working on a preliminary Lasagne implementation of this in the lab, which handles both the 2D and 4D case (and indeed any other dimensionality) but is much more compact. I'm working on this with other people so I'll talk to them and see if they're okay with sharing this code at this point.

Do we want to overload the `deterministic` keyword argument for this? Maybe there should be a different keyword argument specifically for this, because generating test time predictions involves more than just 'disabling' some non-determinism. Then again, I'm not sure if there are any use cases for turning dropout and batch normalization on/off separately.

I'm not sure if adding two methods `pre_train` and `post_train` is a good idea. If we do need to extend the API to support batch normalization, we'll have to discuss how we can do this in as simple a way as possible (same for `get_additional_updates`).

**Comment by takacsg84** [comment]
Sorry, for the BatchNorm layer pull. That was totally unintended. I wanted to keep that on my folk, and just share the link of the file, because some requested. I`m indeed a git newbie. So is it not possible to just pull some selected files from my fork? Is it all or nothing?

**Comment by benanne** [comment]
I'm sure there is a way, but @f0k is better placed than me to answer that question :)

**Comment by ebenolson** [comment]
There's a good description of the procedure [here](http://stackoverflow.com/questions/25955822/git-cherry-pick-a-single-commit-for-pull-request)

**Comment by f0k** [comment]
> So is it not possible to just pull some selected files from my fork? Is it all or nothing?

A pull request is linked to a branch in your fork. In the PR we have here, you requested to merge your `master` branch into Sander's `master` branch. Thus, whatever you push to your `master` branch will become part of this PR. (Even if it is pushed after you opened the PR; this allows you to fix mistakes / address feedback by reviewers.)

Usually, you would create a separate branch to implement a feature, then send a PR to merge that branch. There's [a nice tutorial by github](https://guides.github.com/introduction/flow/) itself.

Now that this PR is already for `master`, the easiest would be to change the contents of your `master` branch to what you wanted in the PR, otherwise you'd need to file a new PR. Before you do that, make sure your current work is committed locally, otherwise you cannot freely switch branches. Then do:
```bash
# check again that everything is committed
git status
# if there are local changes, you need to commit them (or stash them, but let's not introduce too many concepts at once)
# create a new branch for your batch normalization experimental code
git checkout -b batch-normalization
# switch back to master
git checkout master
# reset master such that it only contains the #143 fix
# warning: any uncommitted local changes will be lost, so be sure you committed everything
git reset --hard ab65dbc
# push to the github server (this needs a force-push)
git push --force origin master
```
The PR should be in the state we want it then. To continue working on your batch normalization code, just do `git checkout batch-normalization` then.

**Comment by takacsg84** [comment]
@f0k 
Thx for your prompt responses. I followed your cheat sheet and I think now it is in the state that you requested, though I`m not sure how to check it.

@benanne
>From a quick look at the code (I haven't looked at in detail yet) I feel like it could be a lot shorter.

Well I "picked up" python like 3 weeks ago, so probably it feels more like a c++ code, but I`m not just here to contribute but to learn also. I`m not sure though that it is possible to generalize for all input shapes, so i'm not sure if any essential steps can be skipped.
>Do we want to overload the deterministic keyword argument for this?

Overloading might not be a good idea, but in this case actually does what the literal meaning for the word is :)
>I'm not sure if adding two methods pre_train and post_train is a good idea.

This is not necessary at all. This is just my personal fetishism. It is possible (and also easy) to work with exponential moving average also for the cost of not being 100% deterministic

**Comment by f0k** [comment]
> I followed your cheat sheet and I think now it is in the state that you requested, though I`m not sure how to check it.

Yes, it is, you can see so on the "Files changed" tab. The travis build has failed for some reason, is there a way to just re-run it? @dnouri?

Besides, we should have a test for your additions (a test that would have failed without your changes).

**Comment by dnouri** [comment]
@f0k: @benanne should see a "Restart Build" button at the top right of this page that should do the job: https://travis-ci.org/benanne/Lasagne/builds/51630399
I haven't ever used it myself though, so not sure if results will be propagated back to Github.

**Comment by benanne** [comment]
I restarted it, we'll see :)

**Comment by f0k** [comment]
It passed! I'd still vote for merging #135 first, then rebasing and adding tests.

**Comment by benanne** [comment]
Agreed, #135 is the older PR (and bigger).

**Comment by benanne** [comment]
#135 was just merged, do you think you could rebase? No worries if you don't have time for this immediately, it's a small change so I could take care of it if necessary.

**Comment by f0k** [comment]
Because you probably don't know what rebasing means, it's:
```
git fetch upstream  # assuming "upstream" is benanne's fork
git rebase upstream/master
# now fix conflicts if needed and do "git add" to mark them resolved
git diff upstream/master # check what your would PR propose now (optional)
git push --force origin master  # to update the PR
```
If that's too complicated, leave it to @benanne. It's just so complicated because both #135 and you did some pep8 cleanup along with your actual changes, causing the PRs to affect the same lines of codes (and thus conflict with each other).

**Comment by takacsg84** [comment]
@f0k thx for the instructions. It wasn't that complicated
>We should probably pep8 everything once to avoid this in future.

I think that would be a good idea, especially since many have autoformat tools.

**Comment by benanne** [comment]
There was a lengthy discussion before about whether we want to do this (#21), which never really lead to a conclusion. But I think most people were in support of enforcing PEP8. The main concern was the 80 character line limit, which I personally find a little extreme, but I guess a lot of people like it that way.
